pilot/pkg/bootstrap/server.go

```diff
@@ -407,34 +374,54 @@ func (s *Server) initMesh(args *PilotArgs) error {
 	return nil
 }
 
-// initMeshNetworks creates the meshNetworks in the pilotConfig from the input arguments.
+// initMeshNetworks loads the mesh networks configuration from the file provided
+// in the args and add a watcher for changes in this file.
 func (s *Server) initMeshNetworks(args *PilotArgs) error {
-	// If a config file was specified, use it.
-	if args.MeshNetworks != nil {
-		s.meshNetworks = args.MeshNetworks
+	if args.NetworksConfigFile == "" {
+		log.Info("mesh networks configuration not provided")
 		return nil
 	}
+
 	var meshNetworks *meshconfig.MeshNetworks
 	var err error
 
-	if args.NetworksConfigFile != "" {
-		meshNetworks, err = cmd.ReadMeshNetworksConfig(args.NetworksConfigFile)
-		if err != nil {
-			log.Warnf("failed to read mesh networks configuration, using default: %v", err)
-		}
+	meshNetworks, err = cmd.ReadMeshNetworksConfig(args.NetworksConfigFile)
+	if err != nil {
+		log.Warnf("failed to read mesh networks configuration from %q. using default.", args.NetworksConfigFile)
+		return nil
 	}
+	log.Infof("mesh networks configuration %s", spew.Sdump(meshNetworks))
+	s.meshNetworks = meshNetworks
 
-	if meshNetworks == nil {
-		// Config file either wasn't specified or failed to load - use a default
-		if _, meshNetworks, err = GetMeshNetworks(s.kubeClient, kube.IstioNamespace, kube.IstioNetworksConfigMap); err != nil {
-			log.Warnf("failed to read mesh networks configuration: %v", err)
-			return err
+	// Watch the config file for changes and reload if it got modified
+	fileWatcher := filewatcher.NewWatcher()
+	fileWatcher.Add(args.NetworksConfigFile)
+	go func() {
+		var timerC <-chan time.Time
+		for {
+			select {
+			case <-timerC:
+				timerC = nil
+				// Reload the config file
+				meshNetworks, err = cmd.ReadMeshNetworksConfig(args.NetworksConfigFile)
+				if err != nil {
+					log.Warnf("failed to read mesh networks configuration from %q", args.NetworksConfigFile)
+					continue
+				}
+				if reflect.DeepEqual(meshNetworks, s.meshNetworks) {
+					log.Infof("mesh networks configurtion file updated to: %s", spew.Sdump(meshNetworks))
+					s.meshNetworks = meshNetworks
+					s.EnvoyXdsServer.ConfigUpdate(true)
```

Comment:
This is not a config update.. IOW, this shouldn't trigger LDS/RDS/CDS updates. (well may be CDS). But it should trigger EDS recomputation. So may be create a separate callback for this as well as global mesh config update?

Reply:
@rshriram I updated to do a deep ConfigUpdate on both mesh and networks config changes (only if they new one is different than the old).
Because config updates are rare I think it's fine to purge all?

Reply:
yeah